### ISSUE TITLE
Tiddler title handling should trim whitespace

### ISSUE THREAD

**Issue by pmario** [Jermolene/TiddlyWiki5#751]
The following code creates 3 different tiddler links.
IMO they should all link to the same tiddler

```
\define link(title:"no title specified!")
<$link to="$title$">$title$</$link>
\end

<<link "xxx asdf">>
<<link "xxx asdf ">>
<<link " xxx asdf ">>
```

**Comment by ng110** [comment]
Hi,

I came across this old discussion while looking to see if there were any previous discussions relating to a trailing space problem that I have encountered. I'll come back to my problem in a moment, but I just wanted to first weigh in on this discussion, since I disagree with the consensus so far expressed regarding desirability of having extra spaces matter.

While trimming out whitespace during rendering is fine in most cases, I think that deleting text characters that the user has entered is (almost) always wrong.  I can't see any reason why the user should be prevented from using extra spaces if they choose to.  Yes you have to be careful when linking to them, but that's not different from the need to get the spelling right.  Currently the behaviour is inconsistent because when editing a tiddler, extra whitespace is removed from the start or end of a title but not from the middle.  I see no problem at all with just allowing the title to stay as the user has written it.  I would however argue that to make this more obvious once it has been done, then links to the tiddler should be rendered with all the whitespace exactly as-is, so the distinction isn't hidden.

The other argument is that while current ways of creating tiddlers can force particular handling of whitespace, there is no guarantee that all future ways of doing so will be able to (or will want to), so TW ought to be able to handle cases when extra whitespace exists in a consistent way.  This relates to my problem.  I have written a Python module to interface to a TW file (cf https://groups.google.com/forum/#!searchin/TiddlyWiki/python/tiddlywiki/BBM5IyiLGsI/I2QWCIl-KsEJ), and this can put arbitrary text into a tiddler title, or any other field.  In one of my wikis, I created lots of tiddlers from an external data source using this code, and that external data happened to have a number of items where the title information had trailing spaces.  Although TW can't normally create such titles, it does allow them to exist, and treats them as different from an equivalent tiddler without the trailing space: that is, both can exist concurrently, and links to them need to accurately include or exclude the space.  Because of this, I would like to remove the trailing spaces where they exist, but when I edit one of these tiddlers and remove the space and save the change, TW doesn't recognise that there has been a change to the title, so instead of overwriting the old tiddler, it makes a duplicate.

Obviously I can then delete the unwanted one, and I could alternatively use my Python code to remove the trailing spaces, so this isn't really a significant problem for me, but it seems like incorrect behaviour that TW automatically assumes that there was no extra whitespace when determining whether the title has changed.

The discussion so far notes the difficulty of backward compatibility in making spaces not matter, but I don't see that such an issue would exist in making it such that spaces always do matter.

Cheers,

Neil.

**Comment by Jermolene** [comment]
Hi @ng110 the trimming of whitespace from the tiddler title is done by the navigator widget during processing the `tm-save-tiddler` message. There's no equivalent processing elsewhere when tiddlers are created or modified; internally, titles with trailing or leading spaces should just work in the expected way.

**Comment by pmario** [comment]
I still think, trailing spaces in tiddler titles should be consistently removed, since browser vendors, decided not to render them! IMO for a good reason. Invisible characters only create edge cases and create support or change requests.

@ng110 
If you programmatically create 2 tiddlers eg: `asdf` and `asdf `. TW core renders them in the way you suggest. If you inspect the DOM, you'll see, that the second tiddler title heading element contains the trailing space. ... **but the browsers don't render it**, imo for a good reason. 

In TW, if you open both tiddlers `asdf` in view mode, `asdf ` in edit mode and save. TW doesn't duplicate the tiddler. It overwrites `asdf` without any warning. ... **Which imo is a problem, because it means data loss.** It shows the same tiddler 2 times in view mode. 

code for dev console to test it: 

```
$tw.wiki.addTiddler(new $tw.Tiddler({title: "asdf", text: "asdf"}, $tw.wiki.getModificationFields()));
$tw.wiki.addTiddler(new $tw.Tiddler({title: "asdf ", text: "blabla"}, $tw.wiki.getModificationFields()));
```

If you rename / save `asdf ` the text will be `blabla`

As you point out, it is inconsistent, but imo the only way to fix it, is to make the behaviour consistent with what browsers do anyway: They don't show the trailing spaces.

**Comment by Jermolene** [comment]
Hi @pmario I don't think trying to match browser whitespace semantics is a good goal: for example, HTML also coalesces adjacent spaces, which I don't think we'd want.

**Comment by pmario** [comment]
>Hi @pmario I don't think trying to match browser whitespace semantics is a good goal: for example, HTML also coalesces adjacent spaces, which I don't think we'd want.

We may agree, that we disagree. I think, most of the time we want it. But you are right. What ever is the default, there will be someone who has a problem with it :) ... 

@ng110 

I think we have to live with the status quo and I'm ok with it. So I'd like to close the issue, since the OP won't happen. 

IMO in your proposal you want to allow leading and trailing spaces in the TW EditTemplates, which is the opposite of the OP. So feel free to open a new issue but I'd vote -1 on it. sry.

**Comment by felixhayashi** [comment]
TL;DR

I would never allow trailing spaces in a title. All websites I know remove spaces in form entries (except pasword fields).

**Comment by Jermolene** [comment]
> I would never allow trailing spaces in a title

But what does that mean, exactly? Some possibilities:

1. `wiki.addTiddler()` raises an error for titles with trailing spaces (and maybe multiple contiguous spaces)
2. `wiki.addTiddler()` massages such titles into a legal form, with new semantics for returning the massaged title
3. The user interface tries to make it hard for people to accidentally create tiddlers with trailing spaces

I really don't think that any of those are viable except (3), which is what we try to do now.

**Comment by ng110** [comment]
> All websites I know remove spaces in form entries

But TW is not a website.  It is a data management application and should therefore record and store data as provided.  TW has no business telling me that I can't store data with trailing spaces if I like.  Seems antithetical to the TW philosophy of having control of our own data.

> I would never allow trailing spaces in a title
> 
> But what does that mean, exactly? Some possibilities:
> 
>     wiki.addTiddler() raises an error for titles with trailing spaces (and maybe multiple contiguous spaces)
>     wiki.addTiddler() massages such titles into a legal form, with new semantics for returning the massaged title
>     The user interface tries to make it hard for people to accidentally create tiddlers with trailing spaces

Surely we have to distinguish between what is stored and what is rendered?  I have no problem stripping out whitespace for rendering, but leave the underlying data alone.  I don't see any need to force the user to remove the characters.  Arbitrarily removing them creates more inconsistency than leaving them there (as far as I can see).

**Comment by Jermolene** [comment]
>  I have no problem stripping out whitespace for rendering, but leave the underlying data alone. I don't see any need to force the user to remove the characters. Arbitrarily removing them creates more inconsistency than leaving them there (as far as I can see).

At the moment, we don't strip out the whitespace for rendering, but HTML coalesces it for us when it is displayed.

**Comment by ng110** [comment]
> At the moment, we don't strip out the whitespace for rendering, but HTML coalesces it for us when it is displayed.

Ah.  My misunderstanding of what 'rendering' actually is!  What I really meant to say was I don't mind the whitespace being stripped out when formatted for display.

**Comment by felixhayashi** [comment]
@Jermolene 

> wiki.addTiddler() massages such titles into a legal form, with new semantics for returning the massaged title

I would be in favour of stripping the spaces at the time when the constructor is called to create an instance `new $tw.Tiddler()` and the returned object contains the title without trailing spaces.

IMO stripping of spaces should be done seamlessly without any visual warning (popup) or else. Why would somebody want trailing spaces in a title that are not even rendered because the browser thinks it doesn't make sense to display them?! - this one I really don't get.

This only creates problems. What about linking to a tiddler that has a space? What if I use a filter to search for the title and it doesn't work because there is a space in the title? What about title comparissons in a plugin? what about uniqueness of a title? So much danger and confusion can come with this.

@ng110 

>  It is a data management application and should therefore record and store data as provided.

I completely disagree. Why do you think even MySQL removes spaces when inserting a strings in the db?

> Before MySQL 5.0.3, trailing spaces are removed from values when they are stored into a VARCHAR column; this means that the spaces also are absent from retrieved values.

Why do CMS like WordPress, Joomla and other always strip titles? Have a google search and you will see that it is always considered good practice to remove trailing spaces at the very first occurence (which would be at the time `new $tw.Tiddler()` is executed in our case).

> I don't mind the whitespace being stripped out when formatted for display.

When you don't mind whitespaces are stripped off at rendering, why do you need them anyway?

I completeley agree with @pmario saying *I still think, trailing spaces in tiddler titles should be consistently removed*.

-Felix

**Comment by Jermolene** [comment]
Hi @felixhayashi 

> I would be in favour of stripping the spaces at the time when the constructor is called to create an instance `new $tw.Tiddler()` and the returned object contains the title without trailing spaces.

Clearly we can't make that change without breaking backwards compatibility.

> IMO stripping of spaces should be done seamlessly without any visual warning (popup) or else. Why would somebody want trailing spaces in a title that are not even rendered because the browser thinks it doesn't make sense to display them?! - this one I really don't get.

We all agree on the confusion for users that can result from the inadvertent inclusion of trailing/leading spaces. So that's not the disagreement. The disagreement is about the layer at which we deal with the problem: the store vs. the UI.

As I say, the debate is probably academic because the semantics of the store are already established. It is quite intentional that the store just behaves like a dumb hashmap. It makes it much easier to work with programmatically.

**Comment by pmario** [comment]
>>I would be in favour of stripping the spaces at the time when the constructor is called to create an instance new $tw.Tiddler() and the returned object contains the title without trailing spaces.

>Clearly we can't make that change without breaking backwards compatibility.

As I wrote. I'd like to close the issue, since it will be a "wontfix", due to backwards compatibility concerns. 

TW titles have a lot more problems eg: closing braces `}])`, which can only be avoided, if we establish sensible "best practices" in the docs, that allow users to create titles that work. 

Plugin authors, that are hit by the problem can be pointed to this topic. .. done.

**Comment by felixhayashi** [comment]
@Jermolene 

Your last comment clarified a lot. Thanks!

> Clearly we can't make that change without breaking backwards compatibility.

I agree. But I think it wouldn't hurt so much. Anyway, you are right, we can never know.

> We all agree on the confusion for users that can result from the inadvertent inclusion of trailing/leading spaces. So that's not the disagreement.

Ah ok

> It is quite intentional that the store just behaves like a dumb hashmap

Ok, I understood the intention.

**Comment by tobibeer** [comment]
My 5cts: What ui or controlers can't handle has no business to being in the store... and clearly, TiddlyWiki as a store for someone's model comes with constraints. TiddlyWiki's model is tightly coupled with its views and controlers. So, I don't quite see the point to allowing store objects that won't make it to the shelf, intact that is.

**Comment by pmario** [comment]
As Jeremy tagged TW V 5.1.0 there was a very strong commitment to backwards compatibility. 

So as soon as this argument comes, it gets infinitely hard to convince him otherwise. That's the main reason, why I closed the issue. 

Also ng110 has some strong arguments. I also don't want any software to "think for me", so I can understand those arguments too.

**Comment by tobibeer** [comment]
> I think that deleting text characters that the user has entered is (almost) always wrong. I can't see any reason why the user should be prevented from using extra spaces if they choose to.

I disagree with @ng110 on this and I encourage everyone else to do so, too.

There is a clear spec as to what constitutes a valid fieldname — see html spec — and so is or should there be one for tiddler titles, as all that leading or trailing blancs introduce are inconsistencies, like your overwriting issue. Where is the benefit in keeping an inconsistent store?

Ever tried to write a file or folder to your filesystem that starts or ends with blancs? It won't happen and there's a good reason for that.

**Comment by Jermolene** [comment]
> Ever tried to write a file or folder to your filesystem that starts or ends with blancs? It won't happen and there's a good reason for that.

Actually, many filing systems let you create filenames through the API that cannot be manipulated in the user interface. For example, *nix command line systems get problems with files called `*`.

**Comment by tobibeer** [comment]
> As I say, the debate is probably academic because the semantics of the store are already established.

I don't really think there's anything academic to this:

1. Leading or trailing white-space in titles as well as field-names must not be supported.
2. If **any** layer allows it, it is to be considered a bug that needs fixing.

**Comment by felixhayashi** [comment]
Hi @tobibeer 

I agree with you that trimming is the way to go as I see many problems with this. But there may be some programs that rely on the fact the the core doesn't trim. E.g. consider this: js program somehow receives a string with leading/trailing spaces and uses it as tiddler title and later searches for the tiddler and cannot find it. Why? Because the store removed the spaces. However, I find this scenario not realistic enough to be against trimming. But in any case, trimming would break backwards compatibility and I think @Jermolene got a point with being strict here. Not sure what I would do, I think I would still favour trimming :blush:

**Comment by tobibeer** [comment]
>  However, I find this scenario not realistic enough to be against trimming.

Me too. The core spec should clearly outrule whitespace for both field-names as well as the title field. So, any implementation that would rely on titles with spaces needs to realize those are simply not allowed.

**Comment by tobibeer** [comment]
> But in any case, trimming would break backwards compatibility and I think @Jermolene got a point with being strict here. 

@felixhayashi / @Jermolene: Backwards compatible with respect to what though? This term must not be used to shun somewhat unpopular inconsistencies. These things deserve and require addressing.

If behavior goes against sanity or common sense... it is not reliability on faulty behavior that is of need. Instead, a change of behavior begs to be put on the table.

**Comment by tobibeer** [comment]
Fwiw, I would think the core needs to do either one or the other, but not something in the middle.

Either spaces are considered a problem, and so they are removed wherever they are... or they are not. There should not be parts of the implementation that fix something while eventually breaking things on the other end, as in the OP. The key is consistency.

If the simple solution is to actually always allow spaces where they are permissible, even if invisible and that being an unpopular or "bad" choice for some, e.g. for titles, then we should nevertheless allow it.

If the simple solution for another spot is that white space is never persmissible, such as field names, then we must always ensure to sanitize a given string if a user or program attempts to define an otherwise invalid fieldname.

Perhaps, even the decision to only allow html attribte conventions for field names was an unnecessary one that eventually makes many use cases impossible or a lot more difficult to implement, i.e. requiring something like external state tiddler titles with full titles at the end in order to define the desired references that would allow otherwise unsupported naming patterns.

**Comment by pmario** [comment]
>Perhaps, even the decision to only allow html attribte conventions for field names was an unnecessary one ...

That's not true, because of browser quirks. ... Tiddler fields are stored in html attributes, which doesn't survive a "save / reload" roundtrip if we allow the full spec for attributes.

... It would need a rewrite of the mechanism, how we store tiddlers, which was discussed several times. 

But imo that's a resource problem.

**Comment by tobibeer** [comment]
> It would need a rewrite of the mechanism, how we store tiddlers, which was discussed several times.

yes, sure... I'm not suggesting this is something we should be changing... just pointing out that that's the only reason for there to be this kind of constraint in the first place

However, for titles, this constraint does not apply.

**Comment by pmario** [comment]
>However, for titles, this constraint does not apply.

We could add the same warning that we have with `]}` and so on. If we detect trailing whitespace. ... But that doesn't solve the problem if the title is generated programatically.

**Comment by tobibeer** [comment]
> But that doesn't solve the problem if the title is generated programatically.

I don't think there is such a problem. I agree, that visually it's hard to track, but once you open the editor any leading or trailing blanks become pretty obvious.

**Comment by pmario** [comment]
trailing blanks are invisible, except you highlight / select the title field.

**Comment by tobibeer** [comment]
> except you highlight / select the title field

...which is what happens by default when opening edit mode
